### ISSUE TITLE
Custom dataset, training dies after 1 epoch

### ISSUE THREAD

**Issue by damonftl** [open-mmlab/mmdetection3d#1856]
Hi, I've created a custom synthetic dataset and I have managed to get it to load and train for one epoch, after which it crashes every time. This is what is output before the actual crash, so you can see how far it made it:

```
2022-09-20 08:58:49,066 - mmdet - INFO - Epoch [1][50/226]	lr: 4.995e-05, eta: 1:11:12, time: 0.795, data_time: 0.044, memory: 3396, loss_cls: 1.0794, loss_bbox: 2.5220, loss_dir: 0.0790, loss: 3.6804, grad_norm: 92.3445
2022-09-20 08:59:27,583 - mmdet - INFO - Epoch [1][100/226]	lr: 9.990e-05, eta: 1:09:26, time: 0.770, data_time: 0.002, memory: 3396, loss_cls: 1.0202, loss_bbox: 1.8689, loss_dir: 0.0530, loss: 2.9421, grad_norm: 78.9453
2022-09-20 09:00:02,303 - mmdet - INFO - Epoch [1][150/226]	lr: 1.499e-04, eta: 1:06:12, time: 0.694, data_time: 0.002, memory: 3403, loss_cls: 0.9288, loss_bbox: 1.1846, loss_dir: 0.0254, loss: 2.1387, grad_norm: 46.8588
2022-09-20 09:00:40,658 - mmdet - INFO - Epoch [1][200/226]	lr: 1.998e-04, eta: 1:05:52, time: 0.767, data_time: 0.002, memory: 3403, loss_cls: 0.8908, loss_bbox: 0.8952, loss_dir: 0.0164, loss: 1.8024, grad_norm: 31.8395
2022-09-20 09:01:01,023 - mmdet - INFO - Saving checkpoint at 1 epochs
[                                                  ] 0/24, elapsed: 0s, ETA:Traceback (most recent call last):
```
So it gets through that first epoch and saves the checkpoint, and just as it is about to start what I think is the evaluation step, that's when it crashes. After the stack trace, this is the final error output when it stops:

```
  File "/home/damon/anaconda3/envs/openmmlab/lib/python3.8/site-packages/mmcv/ops/voxelize.py", line 71, in forward
    size=(max_voxels, max_points, points.size(1)))
IndexError: Dimension out of range (expected to be in range of [-1, 0], but got 1)
```

What it looks like to me is that the evaluation (or maybe test?) dataset is not loaded correctly for some reason, but I'm not sure. When I print out the data in voxelize.py before line 71, it does suddenly change when not using the training dataset, but I don't know why. I'm glad to provide any configs if it would help. My config files started from the provided kitti and waymo configs, which I edited for my particular data.

**Comment by damonftl** [comment]
An update: I can definitively say the problem is with the evaluation step. If I add this line to my dataset config:
```
evaluation = dict(interval=5, pipeline=eval_pipeline)
```
then it will train for 5 epochs no problem before the "Dimension out of range" error comes up.

Here's my data dict:
```
data = dict(
	samples_per_gpu=1,
	workers_per_gpu=2,
    train=dict(
        type=dataset_type,
        data_root=data_root,
        ann_file=data_root + "training.pkl",
        pipeline=train_pipeline,
        filter_empty_gt=False,
        classes=class_names,
        test_mode=False,
        box_type_3d='LiDAR'),
    val=dict(
        type=dataset_type,
        data_root=data_root,
        ann_file=data_root + "eval.pkl",
        pipeline=eval_pipeline,
        filter_empty_gt=False,
        classes=class_names,
        test_mode=True,
        box_type_3d='LiDAR'),
    test=dict(
        type=dataset_type,
        data_root=data_root,
        ann_file=data_root + "eval.pkl",
        pipeline=test_pipeline,
        filter_empty_gt=False,
        classes=class_names,
        test_mode=True,
        box_type_3d='LiDAR'))
```
Is there anything obviously wrong with that? The val and test data are loaded exactly the same way as the train data. Here's the eval pipeline in case that will help:
```
eval_pipeline = [
    dict(
        type='LoadPointsFromFile',
        coord_type='LIDAR',
        load_dim=7,
        use_dim=3),
    dict(
        type='DefaultFormatBundle3D',
        class_names=class_names,
        with_label=False),
    dict(type='Collect3D', keys=['points'])
]
```

**Comment by damonftl** [comment]
Just another small update on this in the hopes it will help someone to help me figure this out. I've now had exactly the same error with using both a PointPillars model (which is what the above errors are from) and also with the CenterPoint model, which I managed to get training today.

In both models, it is exactly the same problem: training goes for as many epochs as I let it, but as soon as a test or eval pipeline are started, it crashes out with the Dimension out of range error above in voxelize.py. The same exact error also occurs when running tools/test.py with my configurations. As far as I can tell, the problem is not with my custom data because the training step works and model files for each epoch are even saved to the output folder.

Here's my dataset class, maybe that will help too:
```
# dataset settings
dataset_type = 'MyDataset'
data_root = 'data/mydataset/'

class_names = ['object1', 'object2', 'object3']

pc_range = [-10.0, -100.0, -5.0, 500.0, 100.0, 10.0]
input_modality = dict(use_lidar=True, use_camera=False)

train_pipeline = [
    dict(
        type='LoadPointsFromFile',
        coord_type='LIDAR',
        load_dim=7,
        use_dim=3),
    dict(
        type='LoadAnnotations3D',
        with_bbox_3d=True,
        with_label_3d=True),
    dict(type='PointsRangeFilter', point_cloud_range=pc_range),
    dict(type='ObjectRangeFilter', point_cloud_range=pc_range),
    dict(type='ObjectNameFilter', classes=class_names),
    dict(type='PointShuffle'),
    dict(type='DefaultFormatBundle3D', class_names=class_names),
    dict(type='Collect3D', keys=['points', 'gt_bboxes_3d', 'gt_labels_3d'])
]

test_pipeline = [
    dict(
        type='LoadPointsFromFile',
        coord_type='LIDAR',
        load_dim=7,
        use_dim=3),
    dict(
        type='DefaultFormatBundle3D', 
        class_names=class_names, 
        with_label=False),
    dict(type='Collect3D', keys=['points'])
]

eval_pipeline = [
    dict(
        type='LoadPointsFromFile',
        coord_type='LIDAR',
        load_dim=7,
        use_dim=3),
    dict(
        type='DefaultFormatBundle3D',
        class_names=class_names,
        with_label=False),
    dict(type='Collect3D', keys=['points'])
]

data = dict(
	samples_per_gpu=1,
	workers_per_gpu=2,
    train=dict(
        type=dataset_type,
        data_root=data_root,
        ann_file=data_root + "training.pkl",
        pipeline=train_pipeline,
        filter_empty_gt=False,
        classes=class_names,
        test_mode=False,
        box_type_3d='LiDAR'),
    val=dict(
        type=dataset_type,
        data_root=data_root,
        ann_file=data_root + "eval.pkl",
        pipeline=eval_pipeline,
        filter_empty_gt=False,
        classes=class_names,
        test_mode=True,
        box_type_3d='LiDAR'),
    test=dict(
        type=dataset_type,
        data_root=data_root,
        ann_file=data_root + "eval.pkl",
        pipeline=test_pipeline,
        filter_empty_gt=False,
        classes=class_names,
        test_mode=True,
        box_type_3d='LiDAR'))

evaluation = dict(interval=1, pipeline=eval_pipeline)
```

**Comment by wittmeis** [comment]
I am not 100% sure but I think I stumbled on this as well and the reason is that in the training pipeline the data is presented in batches and hence the individual data frames are encapsulated in a list.

This list is missing in your test and eval pipelines and therefore the voxelizer is receiving the data tensor directly. If I am correct, than this ultimately leads to `points` in https://github.com/open-mmlab/mmcv/blob/3b9b9054bf4b5195d57bf6dfc6a60860b0a3cec1/mmcv/ops/voxelize.py#L70 being not a two dimensional tensor as expected.

In the example configs in the repo the test and eval pipelines always contain a `MultiScaleFlipAug3D` transform: https://github.com/open-mmlab/mmdetection3d/blob/76e351a7a6b0b83d7dc729bdfee5f17743c72f76/configs/_base_/datasets/kitti-3d-3class.py#L61

I assume that this fixes the issue by re-encapsulating the data as for the training use-case. But frankly I never checked the code of this transform to validate this assumption. But simply give it a try...

**Comment by damonftl** [comment]
Hi wittmeis, thank you for the response! As it turns out, you are correct. I found out the same thing just as my day was finishing yesterday. That MultiScaleFlipAug3D is absolutely required, or testing and evaluation fail and give the error I posted. I took it out of my pipelines because I didn't want to do it on my data. 

The docs don't mention it's a required field, so I didn't think it would be an issue. In addition, it is NOT required for the training pipeline, which added to my confusion. Training works perfectly fine without it, but not testing or evaluation. My dataset does not contain images, so including a transform that has an "img_scale" field seemed wrong.

In case others come across this, here's my working eval pipeline:
```
eval_pipeline = [
    dict(
        type='LoadPointsFromFile',
        coord_type='LIDAR',
        load_dim=7,
        use_dim=3),
    dict(
        type='MultiScaleFlipAug3D',
        img_scale=(1, 1),
        pts_scale_ratio=1,
        flip=False,
        transforms=[
            dict(
                type='GlobalRotScaleTrans',
                rot_range=[0, 0],
                scale_ratio_range=[1., 1.],
                translation_std=[0, 0, 0]),
            dict(
                type='PointsRangeFilter', point_cloud_range=pc_range),
            dict(
                type='DefaultFormatBundle3D',
                class_names=class_names,
                with_label=False),
            dict(type='Collect3D', keys=['points'])
        ])
]
```

So now evaluation completes, but I still can't complete a training. I'm running into an issue with a key missing in my annotations during the step for calculating accuracy. It's not listed in the docs in the tutorial on pure point clouds, so I don't know exactly what it means. That's a separate issue though so I'll go ahead and close this one.

**Comment by KangHyunwoogi** [comment]
hello, thanks for detailed situation description about error.
I'm in completely same situation with you.. I guess this error is from my dataset which don't include image and calib.
If you have solved this problem, may I know how you solved it? 
thanks.

**Comment by damonftl** [comment]
Hello, I did actually manage to get past that last issue I mentioned about a key missing. Unfortunately it was a few months ago and now I don't remember exactly what I did. I can say for sure that once I did manage to complete full trainings I was never able to get any detection accuracy above 3% anyway, and for unknown reasons. We ended up abandoning MMDetection3D not long after I posted these messages for a different solution using 2D networks. I will be looking at this again soon however, and if I can remember how I solved the key missing problem I'll be sure to post here.
